volatility/framework/plugins/windows/cachedump.py

```diff
@@ -0,0 +1,138 @@
+# This file is Copyright 2020 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+from volatility.framework import interfaces, renderers
+from volatility.framework.configuration import requirements
+from volatility.framework.renderers import format_hints
+from volatility.framework.layers import intel
+from volatility.plugins.windows.registry import hivelist
+from volatility.plugins.windows import hashdump, lsadump, poolscanner
+from Crypto.Hash import HMAC
+from Crypto.Cipher import ARC4, AES
+from struct import unpack
+
+
+class Cachedump(interfaces.plugins.PluginInterface):
+    """Dumps lsa secrets from memory"""
+
+    _version = (1, 0, 0)
+    
+    @classmethod
+    def get_requirements(cls):
+        return [requirements.TranslationLayerRequirement(name = 'primary',
+                                                     description = 'Memory layer for the kernel',
+                                                     architectures = ["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name = "nt_symbols",
+                                                description = "Windows kernel symbols"),
+            requirements.PluginRequirement(name = 'hivelist', plugin = hivelist.HiveList, version = (1, 0, 0)),
+            requirements.PluginRequirement(name = 'lsadump', plugin = lsadump.Lsadump, version = (1, 0, 0))
+            ]
+
+    def get_nlkm(self, sechive, lsakey, is_vista_or_later):
+        return lsadump.Lsadump.get_secret_by_name(sechive, 'NL$KM', lsakey, is_vista_or_later)
+    
+
+    def decrypt_hash(self, edata, nlkm, ch, xp):
+        if xp:
+            hmac_md5 = HMAC.new(nlkm.encode('latin1'), ch)
+            rc4key = hmac_md5.digest()
+            rc4 = ARC4.new(rc4key)
+            data = rc4.encrypt(edata)
+        else:
+            # based on  Based on code from http://lab.mediaservice.net/code/cachedump.rb
+            aes = AES.new(nlkm.encode('latin1')[16:32], AES.MODE_CBC, ch)
```

Comment:
nlkm gets converted to a string by the `Lsadump.get_secret_by_name` and then gets converted back to bytes here.  That's liable to introduce sutble guys.  There shouldn't really be any need to encode or decode unless what you're actually dealing with is a string not a series of bytes, and in cryptography, it's almost always a series of bytes...